### ISSUE TITLE
When do I set Descendant?

### ISSUE THREAD

**Issue by grandrx10** [dilevin/computer-graphics-bounding-volume-hierarchy#100]
Within AABBTree_ray_intersection, I recursively set descendant but I never implement a base case for it. I'm wondering where I can actually implement this base case, or if I even need to (does an object have a base implemented descendant return?)

The reason why I am asking about this is because when running the code, I encountered a seg fault (see below) which I traced to a function that uses the descendant return.

![image](https://github.com/user-attachments/assets/a92993f6-324d-43d4-9b63-ded255397828)
![image](https://github.com/user-attachments/assets/6e40d676-dddf-43ff-adb1-b7841eb48953)

Can I get some clarification on how the descendant is supposed to be used? (I have read the other post about descendant, and I am still unclear on how to do so.)

**Comment by Zhecheng-Wang** [comment]
The base case would be if the ray has intersected with a box.
Then it checks if there is left descendant, it checks if the ray intersects with left descendant (tree). The same goes for right tree.
So, if the left descendent is a leaf, then the `descendant` should be set to the leaf object pointer. Otherwise the `descendant` is just the left tree pointer. The same goes for right tree.

**Comment by grandrx10** [comment]
Can I assume that leaf objects will automatically return themselves as a descendant when I call ray intersect on them?

**Comment by Zhecheng-Wang** [comment]
Yes I believe so. The `ray_intersect` should be overloaded when called on a leaf object.
So the tricky part is how do you know if the returned descendant object is a leaf or not.
